src/main/java/com/faforever/client/fx/BrowserCallback.java

```diff
@@ -0,0 +1,223 @@
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.faforever.client.util.RatingUtil.getGlobalRating;
+import static com.faforever.client.util.RatingUtil.getLeaderboardRating;
+
+@SuppressWarnings("WeakerAccess")
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class BrowserCallback {
+  private final PlatformService platformService;
+  private final UrlPreviewResolver urlPreviewResolver;
+  private final ReplayService replayService;
+  private final EventBus eventBus;
+  private final ExternalReplayInfoGenerator externalReplayInfoGenerator;
+  private final Pattern replayUrlPattern;
+  private final ClanService clanService;
+  private final UiService uiService;
+  private final PlayerService playerService;
+  private final I18n i18n;
+  @VisibleForTesting
+  Popup clanInfoPopup;
+  private Tooltip linkPreviewTooltip;
+  private Popup playerInfoPopup;
+  private double lastMouseX;
+  private double lastMouseY;
+
+  BrowserCallback(PlatformService platformService, ClientProperties clientProperties,
+                  UrlPreviewResolver urlPreviewResolver, ReplayService replayService, EventBus eventBus,
+                  ExternalReplayInfoGenerator externalReplayInfoGenerator, ClanService clanService, UiService uiService, PlayerService playerService, I18n i18n) {
+    this.platformService = platformService;
+    this.urlPreviewResolver = urlPreviewResolver;
+    this.replayService = replayService;
+    this.eventBus = eventBus;
+    this.externalReplayInfoGenerator = externalReplayInfoGenerator;
+    this.clanService = clanService;
+    this.uiService = uiService;
+    this.playerService = playerService;
+    this.i18n = i18n;
+
+    String urlFormat = clientProperties.getVault().getReplayDownloadUrlFormat();
+    String[] splittedFormat = urlFormat.split("%s");
+    replayUrlPattern = Pattern.compile(Pattern.quote(splittedFormat[0]) + "(\\d+)" + Pattern.compile(splittedFormat.length == 2 ? splittedFormat[1] : ""));
+  }
+
+  /**
+   * Called from JavaScript when user clicked a URL.
+   */
+  public void openUrl(String url) {
+    Matcher replayUrlMatcher = replayUrlPattern.matcher(url);
+    if (!replayUrlMatcher.matches()) {
+      platformService.showDocument(url);
+      return;
+    }
+
+    String replayId = replayUrlMatcher.group(1);
+
+    replayService.findById(Integer.parseInt(replayId))
+        .thenAccept(replay -> Platform.runLater(() -> externalReplayInfoGenerator.showExternalReplayInfo(replay, replayId)));
+  }
+
+  /**
+   * Called from JavaScript when user clicks on user name in chat
+   */
+  public void openPrivateMessageTab(String username) {
+    eventBus.post(new InitiatePrivateChatEvent(username));
+  }
+
+  /**
+   * Called from JavaScript when user no longer hovers over an URL.
+   */
+  public void hideUrlPreview() {
+    if (linkPreviewTooltip != null) {
```

Comment:
50 lines below you did a if (x == nulll) { return; } statement for one line inside the if block. Why not do it here aswell?